src/ec/eddsa.rs

```diff
@@ -248,6 +250,105 @@ extern  {
     fn GFp_x25519_sc_reduce(s: &mut UnreducedScalar);
 }
 
+/// Invert a field element using Fermat's Little Theorem.
+///
+/// We are in the situation where we have an input element |a| from the finite
+/// field |F_p|, where |p| is the prime 2^255 - 19. The idea of this algorithm
+/// is to find the inverse of |a| using Fermat's Little Theorem, which implies
+/// the identity: a**(p - 2) == a**-1 (mod p).
+///
+/// Since |p| is prime, we have that |a| and |p| are always coprime as
+/// integers, so the above equation holds for any |a|.
+///
+/// Note that we do not explicitly guard against the case |a| == 0, even though
+/// it has no multiplicative inverse.
+///
+/// This implementation of the modular exponentiation is ultimately derived
+/// from Daniel J. Bernstein's reference implementation in qhasm, via
+/// BoringSSL, and referenced in [Curve15519]. The overall strategy is to
+/// square and multiply against the fixed base |a| to double and add to the
+/// exponent, reusing intermediate computations as we are able.
+///
+/// [Curve25519]: https://cr.yp.to/ecdh.html#curve25519-paper
+#[doc(hidden)]
+#[no_mangle]
+pub unsafe extern fn eddsa_elem_invert(out: &mut Elem, a: &Elem) {
+    // Sets |out| to be |a| * |b| (mod p).
+    fn mul(out: &mut Elem, a: &Elem, b: &Elem) {
+        unsafe { GFp_fe_mul(out.as_mut_ptr(), a.as_ptr(), b.as_ptr()); }
+    }
+
+    // Sets |a| to be |a| * |b| (mod p).
+    fn mul_in_place(a: &mut Elem, b: &Elem) {
+        unsafe { GFp_fe_mul(a.as_mut_ptr(), a.as_ptr(), b.as_ptr()); }
+    }
+
+    // Sets |out| to be |a|**2 (mod p).
+    fn sqr(out: &mut Elem, a: &Elem) {
+        unsafe { GFp_fe_sq(out.as_mut_ptr(), a.as_ptr()); }
+    }
+
+    // Sets |a| to be |a|**2 (mod p).
+    fn sqr_in_place(a: &mut Elem) {
+        unsafe { GFp_fe_sq(a.as_mut_ptr(), a.as_ptr()); }
+    }
+
+    let mut t0 = [0i32; ELEM_LIMBS];
+    let mut t1 = [0i32; ELEM_LIMBS];
+    let mut t2 = [0i32; ELEM_LIMBS];
+    let mut t3 = [0i32; ELEM_LIMBS];
+
+    sqr(&mut t0, a);
+    sqr(&mut t1, &t0);
+    for _ in 1..2 {
+        sqr_in_place(&mut t1);
+    }
+    mul_in_place(&mut t1, a);
+    mul_in_place(&mut t0, &t1);
+    sqr(&mut t2, &t0);
+    mul_in_place(&mut t1, &t2);
+    sqr(&mut t2, &t1);
+    for _ in 1..5 {
+        sqr_in_place(&mut t2);
+    }
+    mul_in_place(&mut t1, &t2);
+    sqr(&mut t2, &t1);
+    for _ in 1..10 {
+        sqr_in_place(&mut t2);
+    }
+    mul_in_place(&mut t2, &t1);
+    sqr(&mut t3, &t2);
+    for _ in 1..20 {
+        sqr_in_place(&mut t3);
+    }
+    mul_in_place(&mut t2, &t3);
+    sqr_in_place(&mut t2);
+    for _ in 1..10 {
+        sqr_in_place(&mut t2);
+    }
+    mul_in_place(&mut t1, &t2);
+    sqr(&mut t2, &t1);
+    for _ in 1..50 {
+        sqr_in_place(&mut t2);
+    }
+    mul_in_place(&mut t2, &t1);
+    sqr(&mut t3, &t2);
+    for _ in 1..100 {
+        sqr_in_place(&mut t3);
+    }
+    mul_in_place(&mut t2, &t3);
+    sqr_in_place(&mut t2);
```

Comment:
Why square `t2` once here, and then 49 times below, instead of just 50 times below?